# Integrated_Molecular-Phenotypic_Profiling_of_Stembryos
This repository is a collection of all the code used to perform the analysis for the manuscript: 

[Integrated Molecular-Phenotypic Profiling Reveals Metabolic Control of Morphological Variation in Stembryos](https://doi.org/10.1101/2023.12.04.569921 )

authored by:
- Alba Villaronga Luque 1,2,10, 
- Ryan Savill 1,2,10, 
- Natalia L처pez-Anguita 3,7,11,
- Adriano Bolondi 4,11,
- Sumit Garai 1,5,8,11,
- Seher Ipek Gassaloglu 1,3,9,
- Aayush Poddar 1,
- Aydan Bulut-Karslioglu 3,
- Jesse V Veenvliet 1,5,6,12,*

1 Stembryogenesis Lab, Max Planck Institute of Molecular Cell Biology and Genetics, Dresden, Germany

2 Faculty of Biology, Technische Universit채t Dresden, Dresden, Germany

3 Stem Cell Chromatin Lab,  Dept.  of Genome Regulation, Max Planck Institute for Molecular Genetics, Germany

4 Department of Genome Regulation, Max Planck Institute for Molecular Genetics, Germany

5 Cluster of Excellence Physics of Life, Technische Universit채t Dresden, Dresden, Germany

6 Center for Systems Biology Dresden, Dresden, Germany

7 Present address: Max Delbr체ck Center for Molecular Medicine (MDC), Berlin, Germany

8 Present address: The Francis Crick Institute, London, United Kingdom

9 Present address: Department of Genetics, Yale School of Medicine, Yale University, New Haven, CT, USA

10 These authors contributed equally to this work and should be considered shared first authors

11 These authors contributed equally to this work and should be considered shared second authors

12 Lead contact

DOI: 

## Abstract

Mammalian stem-cell-based models of embryo development (stembryos) hold great promise in basic and applied research. However, considerable phenotypic variation despite identical culture conditions limits their potential. The biological processes underlying this seemingly stochastic variation are poorly understood. Here, we investigate the roots of this phenotypic variation by intersecting transcriptomic states and morphological history of individual stembryos across stages modeling post-implantation and early organogenesis. Through machine learning and integration of time-resolved single-cell RNA-sequencing with imaging-based quantitative phenotypic profiling, we identify early features predictive of the phenotypic end-state. Leveraging this predictive power revealed that early imbalance of oxidative phosphorylation and glycolysis results in aberrant morphology and a neural lineage bias that can be corrected by metabolic interventions. Collectively, our work establishes divergent metabolic states as drivers of phenotypic variation, and offers a broadly applicable framework to chart and predict phenotypic variation in organoid systems. The strategy can be leveraged to identify and control underlying biological processes, ultimately increasing the reproducibility of in vitro systems.
